examples/contrib/cifar10/main.py

```diff
@@ -18,20 +18,73 @@
 from ignite.contrib.handlers.tensorboard_logger import OutputHandler, OptimizerParamsHandler, GradsHistHandler
 from ignite.contrib.handlers import PiecewiseLinear
 
+import ignite.distributed as ignite_dist
+
 import utils
 
 
-def run(output_path, config):
+@ignite_dist.auto_distributed(backend="gloo", device="cpu")
+def run(config):
+
+    # distributed = dist.is_available() and dist.is_initialized()
+    # rank = dist.get_rank() if distributed else 0
+
+    distributed = ignite_dist.get_world_size() > 0
+    rank = ignite_dist.get_rank()
+
+    if distributed:
+        # let each node print the info
+        if ignite_dist.get_local_rank() == 0:
+            print("\nDistributed setting:")
+            print("\tbackend: {}".format(dist.get_backend()))
+            print("\tworld size: {}".format(dist.get_world_size()))
+            print("\trank: {}".format(dist.get_rank()))
+            print("\n")
 
-    distributed = dist.is_available() and dist.is_initialized()
-    rank = dist.get_rank() if distributed else 0
+    output_path = None
+    # let each node print the info
+    if ignite_dist.get_local_rank() == 0:
+        print("Train {} on CIFAR10".format(config["model"]))
+        print("- PyTorch version: {}".format(torch.__version__))
+        print("- Ignite version: {}".format(ignite.__version__))
+        print("- CUDA version: {}".format(torch.version.cuda))
+
+        print("\n")
+        print("Configuration:")
+        for key, value in config.items():
+            print("\t{}: {}".format(key, value))
+        print("\n")
+
+        # create log directory only by 1 node
+        if (not distributed) or (dist.get_rank() == 0):
+            from datetime import datetime
+
+            now = datetime.now().strftime("%Y%m%d-%H%M%S")
+            gpu_conf = "-single-gpu"
+            if distributed:
+                ngpus_per_node = torch.cuda.device_count()
+                nnodes = dist.get_world_size() // ngpus_per_node
+                gpu_conf = "-distributed-{}nodes-{}gpus".format(nnodes, ngpus_per_node)
+
+            output_path = Path(config["output_path"]) / "{}{}".format(now, gpu_conf)
+            if not output_path.exists():
+                output_path.mkdir(parents=True)
+            output_path = output_path.as_posix()
+            print("Output path: {}".format(output_path))
+
+    # show the distributed config
+    ignite_dist.show_config()
 
     manual_seed(config["seed"] + rank)
 
     # Setup dataflow, model, optimizer, criterion
     train_loader, test_loader = utils.get_dataflow(config, distributed)
     model, optimizer = utils.get_model_optimizer(config, distributed)
-    criterion = nn.CrossEntropyLoss().to(utils.device)
+
+    # TODO check for tpu...
+    device = ignite_dist.get_device()
```

Comment:
Shouldn't this be configured by user ?

Reply:
IMO device in ignite_dist should be automatically used by handlers rather than an argument forwarded to handler by user.